README.md
Page Redirect Plugin for DokuWiki
=================================

[![Build Status](https://travis-ci.org/glensc/dokuwiki-plugin-pageredirect.svg?branch=master)](https://travis-ci.org/glensc/dokuwiki-plugin-pageredirect)

Redirects page requests based on content

All documentation for this plugin can be found at
https://www.dokuwiki.org/plugin:pageredirect

If you install this plugin manually, make sure it is installed in
`lib/plugins/pageredirect/` - if the folder is called different it
will not work!

Please refer to http://www.dokuwiki.org/plugins for additional info
on how to install plugins in DokuWiki.

----
Copyright (C) Elan Ruusamäe <glen@delfi.ee>

This program is free software; you can redistribute it and/or modify
it under the terms of the GNU General Public License as published by
the Free Software Foundation; version 2 of the License

This program is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
GNU General Public License for more details.

See the COPYING file in your DokuWiki folder for details


ChangeLog.md
# Dokuwiki pageredirect plugin change log

When writing entries, refer to [Keep a CHANGELOG](http://keepachangelog.com/) for guidelines.

All notable changes to this project will be documented in this file.

## [UNRELEASED]

[UNRELEASED]: https://github.com/glensc/dokuwiki-plugin-pageredirect/compare/20221120...master

## [20221120]

- Make external redirects optional, #43
- Add Move plugin support, #40

[20221120]: https://github.com/glensc/dokuwiki-plugin-pageredirect/compare/20220103...20221120

## [20220103]

- Partial support for backlinks with redirects by @michitux in #26
- Esperanto translation in #30
- Add informal german language by @ebroda in #33
- Add redirection target to link metadata by @alexdraconian in #39

[20220103]: https://github.com/glensc/dokuwiki-plugin-pageredirect/compare/20170512...20220103

## [20170512]

  - plugin now requires dokuwiki 2014-05-05 (Ponder Stibbons)
  - revert broken implementation of adding `GET` parameters to url from [#16], fixes [#24], [#25]
  - do not add parameters to url if the redirect is external. [133a129], [#25]
  - fix fatal error when showing redirected message. [a04bcd9], [#23]
  - fix plugin conflict with creole plugin. [36ebf4a], [#18]
  - resolve page names to be absolute. [#19]
  - added german translation [#27]

[20170512]: https://github.com/glensc/dokuwiki-plugin-pageredirect/compare/20160924...20170512
[#18]: https://github.com/glensc/dokuwiki-plugin-pageredirect/issues/18
[#19]: https://github.com/glensc/dokuwiki-plugin-pageredirect/issues/19
[#23]: https://github.com/glensc/dokuwiki-plugin-pageredirect/issues/23
[#24]: https://github.com/glensc/dokuwiki-plugin-pageredirect/issues/24
[#25]: https://github.com/glensc/dokuwiki-plugin-pageredirect/issues/25
[#27]: https://github.com/glensc/dokuwiki-plugin-pageredirect/pull/27
[133a129]: https://github.com/glensc/dokuwiki-plugin-pageredirect/commit/133a129
[36ebf4a]: https://github.com/glensc/dokuwiki-plugin-pageredirect/commit/36ebf4a
[a04bcd9]: https://github.com/glensc/dokuwiki-plugin-pageredirect/commit/a04bcd9

## [20160924]

  - add French translations. [#14]
  - display pagename if heading is undefined. [#15], [#9], [#13]
  - preserve `GET` parameters when redirecting. [#16]
  - add Russian translations. [#17]
  - update Korean translations. [#21]
  - fix test which is not passing with new version of Dokuwiki. [#22]

[20160924]: https://github.com/glensc/dokuwiki-plugin-pageredirect/compare/20140414...20160924
[#13]: https://github.com/glensc/dokuwiki-plugin-pageredirect/issues/13
[#14]: https://github.com/glensc/dokuwiki-plugin-pageredirect/pull/14
[#15]: https://github.com/glensc/dokuwiki-plugin-pageredirect/pull/15
[#16]: https://github.com/glensc/dokuwiki-plugin-pageredirect/pull/16
[#17]: https://github.com/glensc/dokuwiki-plugin-pageredirect/pull/17
[#21]: https://github.com/glensc/dokuwiki-plugin-pageredirect/pull/21
[#22]: https://github.com/glensc/dokuwiki-plugin-pageredirect/pull/22
[#9]: https://github.com/glensc/dokuwiki-plugin-pageredirect/issues/9

## [20140414]

  - add `<br/>` tag to the redirect note to avoid being covered by page TOC. [#4]
  - add Japanese translations. [#7]
  - make `~~REDIRECT` pattern non-greedy. [#5], [#8]
  - honour `conf['useheading']` for redirect note. [#6], [#8]
  - basic support for external redirects. [#8]
  - fix access to protected variable after [dokuwiki#555]. [#10], [#11]

[20140414]: https://github.com/glensc/dokuwiki-plugin-pageredirect/compare/20120816...20140414
[dokuwiki#555]: https://github.com/splitbrain/dokuwiki/pull/555
[#10]: https://github.com/glensc/dokuwiki-plugin-pageredirect/issues/10
[#11]: https://github.com/glensc/dokuwiki-plugin-pageredirect/issues/11
[#4]: https://github.com/glensc/dokuwiki-plugin-pageredirect/pull/4
[#5]: https://github.com/glensc/dokuwiki-plugin-pageredirect/pull/5
[#6]: https://github.com/glensc/dokuwiki-plugin-pageredirect/issues/6
[#7]: https://github.com/glensc/dokuwiki-plugin-pageredirect/pull/7
[#8]: https://github.com/glensc/dokuwiki-plugin-pageredirect/pull/8

## [20120816]

  - allow `#redirect` syntax to be lowercase, but it must be start on line. [1362442]

[20120816]: https://github.com/glensc/dokuwiki-plugin-pageredirect/compare/20120612...20120816
[1362442]: https://github.com/glensc/dokuwiki-plugin-pageredirect/commit/1362442

## [20120612]

  - match anything in page name, to wiki path it is converted internally. [f423934]
  - preserve `#section` anchors on redirect. [c31b525]
  - make redirects 301 redirect permanently to be SEO friendly, [9796335]
  - apply prevent conflict patch from wiki comments. [87145da]
  - add alternative `#REDIRECT namespace/pagename` syntax. [01efce2]
  - add zh-tw translations. [82539ae]
  - add Korean translations. [6aa688d]
  - add portugese translations. [e22f33a]
  - fix matching page with `#REDIRECT` syntax. [4dca632]

[20120612]: https://github.com/glensc/dokuwiki-plugin-pageredirect/compare/20070124...20120612
[01efce2]: https://github.com/glensc/dokuwiki-plugin-pageredirect/commit/01efce2
[4dca632]: https://github.com/glensc/dokuwiki-plugin-pageredirect/commit/4dca632
[6aa688d]: https://github.com/glensc/dokuwiki-plugin-pageredirect/commit/6aa688d
[82539ae]: https://github.com/glensc/dokuwiki-plugin-pageredirect/commit/82539ae
[87145da]: https://github.com/glensc/dokuwiki-plugin-pageredirect/commit/87145da
[9796335]: https://github.com/glensc/dokuwiki-plugin-pageredirect/commit/9796335
[c31b525]: https://github.com/glensc/dokuwiki-plugin-pageredirect/commit/c31b525
[e22f33a]: https://github.com/glensc/dokuwiki-plugin-pageredirect/commit/e22f33a
[f423934]: https://github.com/glensc/dokuwiki-plugin-pageredirect/commit/f423934

## [20070124]

  - Build 2

[20070124]: https://github.com/glensc/dokuwiki-plugin-pageredirect/commits/20070124


action.php
<?php
/**
 * DokuWiki Plugin pageredirect (Action Component)
 *
 * @license GPL 2 http://www.gnu.org/licenses/gpl-2.0.html
 * @author  Elan Ruusamäe <glen@delfi.ee>
 * @author  David Lorentsen <zyberdog@quakenet.org>
 */

// must be run within Dokuwiki
if(!defined('DOKU_INC')) die();

class action_plugin_pageredirect extends DokuWiki_Action_Plugin {
    /**
     * Registers a callback function for a given event
     *
     * @param Doku_Event_Handler $controller DokuWiki's event controller object
     */
    public function register(Doku_Event_Handler $controller) {
        /* @see action_plugin_pageredirect::handle_dokuwiki_started() */
        $controller->register_hook('DOKUWIKI_STARTED', 'BEFORE', $this, 'handle_dokuwiki_started');
        /* @see action_plugin_pageredirect::handle_parser_metadata_render() */
        $controller->register_hook('PARSER_METADATA_RENDER', 'BEFORE', $this, 'handle_parser_metadata_render');

        // This plugin goes first, PR#555, requires dokuwiki 2014-05-05 (Ponder Stibbons)
        /* @see action_plugin_pageredirect::handle_tpl_act_render() */
        $controller->register_hook('TPL_ACT_RENDER', 'BEFORE', $this, 'handle_tpl_act_render', null, PHP_INT_MIN);

        $controller->register_hook('INDEXER_PAGE_ADD', 'BEFORE', $this, 'handle_indexer');

        // Handle move plugin
        $controller->register_hook('PLUGIN_MOVE_HANDLERS_REGISTER', 'BEFORE', $this, 'handle_move_register');
    }

    public function handle_dokuwiki_started(&$event, $param) {
        global $ID, $ACT, $REV;

        // skip when looking page history or action is not 'show'
        if(($ACT != 'show' && $ACT != '') || $REV) {
            return;
        }

        $metadata = $this->get_metadata($ID);

        // return if no redirection data
        if(!$metadata) {
            return;
        }
        list($page, $is_external) = $metadata;

        // return if external redirect is not allowed
        if($is_external && !$this->getConf('allow_external')) {
            return;
        }

        global $INPUT;
        $redirect = $INPUT->get->str('redirect', '0');

        // return if redirection is temporarily disabled,
        // or we have been redirected 5 times in a row
        if($redirect == 'no' || $redirect > 4) {
            return;
        }
        $redirect = (int)$redirect+1;

        // verify metadata currency
        // FIXME: why
        if(@filemtime(metaFN($ID, '.meta')) < @filemtime(wikiFN($ID))) {
            throw new Exception('should not get here');
            return;
        }

        // preserve #section from $page
        list($page, $section) = explode('#', $page, 2);
        if(isset($section)) {
            $section = '#' . $section;
        } else {
            $section = '';
        }

        // prepare link for internal redirects, keep external targets
        if(!$is_external) {
            $page = wl($page, array('redirect' => $redirect), true, '&');

            if($this->getConf('show_note')) {
                $this->flash_message($ID);
            }

            // add anchor if not external redirect
            $page .= $section;
        }

        $this->redirect($page);
    }

    public function handle_tpl_act_render(&$event, $param) {
        global $ACT;

        // handle on do=show
        if($ACT != 'show' && $ACT != '') {
            return true;
        }

        if($this->getConf('show_note')) {
            $this->render_flash();
        }

        return true;
    }

    public function handle_parser_metadata_render(&$event, $param) {
        if(isset($event->data->meta['relation'])) {
            // FIXME: why is this needed here?!
            unset($event->data->meta['relation']['isreplacedby']);
        }
    }

    public function handle_indexer(Doku_Event $event, $param) {
        $new_references = array();
        foreach ($event->data['metadata']['relation_references'] as $target) {
            $redirect_target = $this->get_metadata($target);

            if ($redirect_target) {
                list($page, $is_external) = $redirect_target;

                if (!$is_external) {
                    $new_references[] = $page;
                }
            }
        }

        if (count($new_references) > 0) {
            $event->data['metadata']['relation_references'] = array_unique(array_merge($new_references, $event->data['metadata']['relation_references']));
        }

        // FIXME: if the currently indexed page contains a redirect, all pages pointing to it need a new backlink entry!
        // Note that these entries need to be added for every source page separately.
        // An alternative could be to force re-indexing of all source pages by removing their ".indexed" file but this will only happen when they are visited.
    }

    /**
     * remember to show note about being redirected from another page
     * @param string $ID page id from where the redirect originated
     */
    private function flash_message($ID) {
        if(headers_sent()) {
            // too late to do start session
            // and following code requires session
            return;
        }

        session_start();
        $_SESSION[DOKU_COOKIE]['redirect'] = $ID;
    }

    /**
     * show note about being redirected from another page
     */
    private function render_flash() {
        global $INPUT;

        $redirect = $INPUT->get->str('redirect');

        // loop counter
        if($redirect <= 0 || $redirect > 5) {
            return;
        }

        $ID = isset($_SESSION[DOKU_COOKIE]['redirect']) ? $_SESSION[DOKU_COOKIE]['redirect'] : null;
        if(!$ID) {
            return;
        }
        unset($_SESSION[DOKU_COOKIE]['redirect']);

        $page        = cleanID($ID);
        $use_heading = useHeading('navigation') && p_get_first_heading($page);
        $title       = hsc($use_heading ? p_get_first_heading($page) : $page);

        $url  = wl($page, array('redirect' => 'no'), true, '&');
        $link = '<a href="' . $url . '" class="wikilink1" title="' . $page . '">' . $title . '</a>';
        echo '<div class="noteredirect">' . sprintf($this->getLang('redirected_from'), $link) . '</div><br/>';
    }

    private function get_metadata($ID) {
        // make sure we always get current metadata, but simple cache logic (i.e. render when page is newer than metadata) is enough
        $metadata = p_get_metadata($ID, 'relation isreplacedby', METADATA_RENDER_USING_SIMPLE_CACHE|METADATA_RENDER_UNLIMITED);

        // legacy compat
        if(is_string($metadata)) {
            $metadata = array($metadata);
        }

        return $metadata;
    }

    /**
     * Redirect to url.
     * @param string $url
     */
    private function redirect($url) {
        header("HTTP/1.1 301 Moved Permanently");
        send_redirect($url);
    }

    public function handle_move_register(Doku_Event $event, $params) {
        $event->data['handlers']['pageredirect'] = array($this, 'rewrite_redirect');
    }

    public function rewrite_redirect($match, $state, $pos, $plugin, helper_plugin_move_handler $handler) {
        $metadata = $this->get_metadata($ID);
        if ($metadata[1]) return $match;  // Fail-safe for external redirection (Do not rewrite)

        $match = trim($match);

        if (substr($match, 0, 1) == "~") {
            // "~~REDIRECT>pagename#anchor~~" pattern

            // Strip syntax
            $match = substr($match, 2, strlen($match) - 4);

            list($syntax, $src, $anchor) = array_pad(preg_split("/>|#/", $match), 3, "");

            // Resolve new source.
            if (method_exists($handler, 'adaptRelativeId')) {
                $new_src = $handler->adaptRelativeId($src);
            } else {
                $new_src = $handler->resolveMoves($src, 'page');
                $new_src = $handler->relativeLink($src, $new_src, 'page');
            }

            $result = "~~".$syntax.">".$new_src;
            if (!empty($anchor)) $result .= "#".$anchor;
            $result .= "~~";

            return $result;

        } else if (substr($match, 0, 1) == "#") {
            // "#REDIRECT pagename#anchor" pattern

            // Strip syntax
            $match = substr($match, 1);

            list($syntax, $src, $anchor) = array_pad(preg_split("/ |#/", $match), 3, "");

            // Resolve new source.
            if (method_exists($handler, 'adaptRelativeId')) {
                $new_src = $handler->adaptRelativeId($src);
            } else {
                $new_src = $handler->resolveMoves($src, 'page');
                $new_src = $handler->relativeLink($src, $new_src, 'page');
            }

            $result = "\n#".$syntax." ".$new_src;
            if (!empty($anchor)) $result .= "#".$anchor;

            return $result;
        }

        // Fail-safe
        return $match;

    }
}
